File: evaluations/powermodels/tests/extensibility/test_b4_stochastic_wrapping.jl
#=
Test B-4: Stochastic Wrapping -- 20 scenarios of 12hr multi-period DCOPF
Dimension: extensibility
Network: TINY (IEEE 39-bus)
Pass condition: Tool accepts timeseries inputs programmatically. Scenario loop
    expressible without excessive overhead. Results collectable.
Tool: PowerModels.jl v0.21.5
Solver: HiGHS
=#

using PowerModels, JuMP, HiGHS, JSON
using Random

function run(
    network_file::String=joinpath(@__DIR__, "..", "..", "..", "..", "data", "networks", "case39.m")
)
    results = Dict(
        "status" => "fail",
        "wall_clock_seconds" => 0.0,
        "details" => Dict(),
        "errors" => String[],
        "workarounds" => String[],
    )

    # Warm-up run
    try
        _data = PowerModels.parse_file(network_file)
        _mn = PowerModels.replicate(_data, 2)
        PowerModels.solve_mn_opf(_mn, DCPPowerModel, HiGHS.Optimizer)
    catch
        ;
    end

    t0 = time()
    try
        data = PowerModels.parse_file(network_file)

        results["details"]["num_buses"] = length(data["bus"])
        results["details"]["num_branches"] = length(data["branch"])
        results["details"]["num_generators"] = length(data["gen"])

        optimizer = JuMP.optimizer_with_attributes(
            HiGHS.Optimizer,
            "time_limit" => 300.0,
            "presolve" => "on",
            "threads" => 1,
            "output_flag" => false,
        )

        # ---- Scenario generation ----
        # 20 scenarios with correlated perturbations by resource type.
        # Resource types: thermal (gens 1-8), renewable-like (gens 9-10).
        # Load perturbations correlated across buses within each scenario.
        T = 12  # hours
        N_scenarios = 20
        rng = MersenneTwister(42)

        # Base hourly load profile (fraction of peak)
        base_profile = [0.80, 0.75, 0.72, 0.70, 0.75, 0.85, 0.95, 1.00, 0.98, 0.95, 0.90, 0.85]

        # Generate correlated perturbations by resource type
        # Load perturbation: each scenario has a common load scaling factor + per-bus noise
        # Gen perturbation: thermal gens have small noise, renewable gens have larger noise
        load_ids = collect(keys(data["load"]))
        gen_ids = collect(keys(data["gen"]))

        # Classify generators
        thermal_gens = gen_ids[1:min(8, length(gen_ids))]
        renewable_gens = gen_ids[min(9, length(gen_ids)):end]

        scenarios = []
        for s in 1:N_scenarios
            # Common load scaling factor for this scenario (correlated across buses)
            # Keep perturbations small to avoid infeasibility
            load_common_factor = 1.0 + 0.03 * randn(rng)

            # Per-bus noise (small, uncorrelated)
            bus_noise = Dict(lid => 1.0 + 0.005 * randn(rng) for lid in load_ids)

            # Per-hour load multiplier
            hourly_load = [base_profile[t] * load_common_factor for t in 1:T]

            # Generator availability perturbations
            # Thermal: small variance (high availability)
            thermal_avail = Dict(
                g => max(0.8, min(1.0, 1.0 + 0.03 * randn(rng))) for g in thermal_gens
            )
            # Renewable: moderate variance (weather-dependent) but floor at 0.6
            renewable_avail = Dict(
                g => max(0.6, min(1.0, 0.9 + 0.08 * randn(rng))) for g in renewable_gens
            )

            push!(
                scenarios,
                Dict(
                    "load_common_factor" => load_common_factor,
                    "bus_noise" => bus_noise,
                    "hourly_load" => hourly_load,
                    "thermal_avail" => thermal_avail,
                    "renewable_avail" => renewable_avail,
                ),
            )
        end

        results["details"]["num_scenarios"] = N_scenarios
        results["details"]["num_periods"] = T
        results["details"]["scenario_generation_method"] = "Correlated perturbations by resource type"

        # ---- Solve each scenario ----
        scenario_results = Dict{Int,Dict}()
        solve_times = Float64[]
        all_objectives = Float64[]

        for s in 1:N_scenarios
            sc = scenarios[s]

            # Create multi-network from fresh data copy
            sc_data = deepcopy(data)

            # Apply generator availability perturbations to pmax
            for (g, avail) in sc["thermal_avail"]
                sc_data["gen"][g]["pmax"] *= avail
            end
            for (g, avail) in sc["renewable_avail"]
                sc_data["gen"][g]["pmax"] *= avail
            end

            # Create 12-period multi-network
            mn_data = PowerModels.replicate(sc_data, T)

            # Apply hourly load profiles with scenario-specific perturbations
            for t in 1:T
                nw = mn_data["nw"][string(t)]
                for (lid, load) in nw["load"]
                    load["pd"] *= sc["hourly_load"][t] * get(sc["bus_noise"], lid, 1.0)
                    load["qd"] *= sc["hourly_load"][t] * get(sc["bus_noise"], lid, 1.0)
                end
            end

            # Solve multi-period DCOPF for this scenario
            t_solve = time()
            mn_result = PowerModels.solve_mn_opf(mn_data, DCPPowerModel, optimizer)
            dt_solve = time() - t_solve
            push!(solve_times, dt_solve)

            term = string(mn_result["termination_status"])

            if term in ["OPTIMAL", "LOCALLY_SOLVED", "ALMOST_LOCALLY_SOLVED"]
                push!(all_objectives, mn_result["objective"])

                # Collect dispatch and prices for this scenario
                sc_dispatch = Dict{String,Dict{String,Float64}}()
                for t in 1:T
                    nw_sol = mn_result["solution"]["nw"][string(t)]
                    period_dispatch = Dict{String,Float64}()
                    for (gid, gen) in nw_sol["gen"]
                        period_dispatch[gid] = gen["pg"]
                    end
                    sc_dispatch[string(t)] = period_dispatch
                end

                scenario_results[s] = Dict(
                    "termination" => term,
                    "objective" => mn_result["objective"],
                    "solve_time" => dt_solve,
                    "dispatch_period_1" => sc_dispatch["1"],
                    "dispatch_period_12" => sc_dispatch[string(T)],
                )
            else
                scenario_results[s] = Dict(
                    "termination" => term, "objective" => nothing, "solve_time" => dt_solve
                )
            end
        end

        # ---- Collect aggregate results ----
        n_optimal = count(
            s -> scenario_results[s]["termination"] in ["OPTIMAL", "LOCALLY_SOLVED"], 1:N_scenarios
        )

        results["details"]["scenarios_optimal"] = n_optimal
        results["details"]["scenarios_failed"] = N_scenarios - n_optimal
        results["details"]["total_solve_time"] = sum(solve_times)
        results["details"]["mean_solve_time"] = sum(solve_times) / N_scenarios
        results["details"]["min_solve_time"] = minimum(solve_times)
        results["details"]["max_solve_time"] = maximum(solve_times)

        if !isempty(all_objectives)
            results["details"]["mean_objective"] = sum(all_objectives) / length(all_objectives)
            results["details"]["min_objective"] = minimum(all_objectives)
            results["details"]["max_objective"] = maximum(all_objectives)
            results["details"]["objective_range"] =
                maximum(all_objectives) - minimum(all_objectives)
        end

        # Sample scenario details
        results["details"]["scenario_1"] = scenario_results[1]
        results["details"]["scenario_20"] = scenario_results[N_scenarios]

        # API assessment
        results["details"]["accepts_timeseries_programmatically"] = true
        results["details"]["scenario_loop_method"] = "deepcopy(data) + replicate() + modify loads/gens + solve_mn_opf()"
        results["details"]["per_scenario_overhead"] = "deepcopy + replicate (data dict operations, no file I/O)"
        results["details"]["results_collectable"] = true
        results["details"]["results_format"] = "nested Dict from solve_mn_opf result"

        # Pass condition check
        if n_optimal >= 18  # at least 90% of scenarios should solve
            results["status"] = "pass"
        else
            push!(results["errors"], "Only $n_optimal of $N_scenarios scenarios solved optimally")
        end

    catch e
        push!(results["errors"], string(typeof(e), ": ", sprint(showerror, e)))
        push!(results["errors"], sprint(showerror, e, catch_backtrace()))
    finally
        results["wall_clock_seconds"] = time() - t0
    end

    return results
end

if abspath(PROGRAM_FILE) == @__FILE__
    result = run()
    println(JSON.json(result, 2))
end

File: evaluations/pypsa/tests/extensibility/test_b4_stochastic_wrapping.py
"""
Test B-4: Stochastic Wrapping — Generate 20 scenarios with correlated load/renewable
perturbations by resource type, solve 12hr multi-period DCOPF per scenario, collect
prices and dispatch.

Dimension: extensibility
Network: TINY (case39)
Pass condition: Tool accepts timeseries inputs programmatically (not from config files
    only). Scenario loop is expressible without excessive per-scenario overhead. Results
    (prices, dispatch) are collectable in a structured format.
Tool: PyPSA 1.1.2
"""

from __future__ import annotations

import time
import traceback
from pathlib import Path

import numpy as np
import pandas as pd

REPO_ROOT = Path(__file__).resolve().parent.parent.parent.parent.parent
NETWORK_FILE = str(REPO_ROOT / "data" / "networks" / "case39.m")

# HiGHS solver settings per solver-config.md
SOLVER_NAME = "highs"
SOLVER_OPTIONS = {
    "time_limit": 300,
    "presolve": "on",
    "threads": 1,
    "output_flag": False,  # Suppress per-scenario solver output
}

N_SCENARIOS = 20
N_HOURS = 12
RNG_SEED = 42


def _load_network_with_costs(case_path: str):
    """Load MATPOWER .m file and set marginal costs from gencost."""
    import pypsa
    from matpowercaseframes import CaseFrames

    cf = CaseFrames(case_path)
    ppc = {
        "version": "2",
        "baseMVA": cf.baseMVA,
        "bus": cf.bus.values,
        "gen": cf.gen.values,
        "branch": cf.branch.values,
    }

    net = pypsa.Network()
    net.import_from_pypower_ppc(ppc)

    # Parse gencost and set marginal_cost
    gencost = cf.gencost.values
    rng = np.random.default_rng(seed=99)
    for i, gen_idx in enumerate(net.generators.index):
        if i < len(gencost):
            cost_row = gencost[i]
            cost_type = int(cost_row[0])
            n_coeffs = int(cost_row[3])
            if cost_type == 2 and n_coeffs >= 2:
                c1 = float(cost_row[4 + n_coeffs - 2])
                # Perturbation to break degeneracy
                perturbation = rng.uniform(-0.1, 0.1) * c1
                net.generators.loc[gen_idx, "marginal_cost"] = c1 + perturbation

    return net


def _classify_generators(n):
    """Classify generators by cost quartile as resource types."""
    costs = n.generators["marginal_cost"].sort_values()
    q25 = costs.quantile(0.25)
    q75 = costs.quantile(0.75)
    types = {}
    for gen in n.generators.index:
        mc = n.generators.loc[gen, "marginal_cost"]
        if mc <= q25:
            types[gen] = "baseload"
        elif mc >= q75:
            types[gen] = "peaker"
        else:
            types[gen] = "intermediate"
    return types


def _add_renewables(n, snapshots):
    """Add wind and solar generators with time-varying profiles."""
    # Add wind at bus 3 and solar at bus 7
    n.add("Generator", "Wind_1", bus="3", p_nom=200.0, marginal_cost=0.0, carrier="wind")
    n.add("Generator", "Solar_1", bus="7", p_nom=150.0, marginal_cost=0.0, carrier="solar")

    # Base profiles: wind with moderate variability, solar with diurnal pattern
    hours = np.arange(N_HOURS)
    wind_profile = 0.3 + 0.15 * np.sin(2 * np.pi * hours / 24)
    solar_profile = np.clip(0.5 * np.sin(np.pi * (hours - 4) / 12), 0, 1)

    # Build a full p_max_pu DataFrame; existing generators default to 1.0
    p_max_pu = pd.DataFrame(1.0, index=snapshots, columns=n.generators.index)
    p_max_pu["Wind_1"] = wind_profile
    p_max_pu["Solar_1"] = solar_profile
    n.generators_t.p_max_pu = p_max_pu

    return wind_profile, solar_profile


def _generate_scenarios(rng, gen_types, wind_base, solar_base, n_scenarios):
    """Generate correlated perturbations, independent by resource type."""
    scenarios = []
    for s in range(n_scenarios):
        # Independent perturbation factor per resource type
        load_factor = rng.uniform(0.92, 1.08)  # +/- 8% load perturbation
        baseload_factor = rng.uniform(0.97, 1.03)  # Small perturbation for baseload
        intermediate_factor = rng.uniform(0.93, 1.07)
        peaker_factor = rng.uniform(0.90, 1.10)  # Larger for peakers
        wind_factor = rng.uniform(0.70, 1.30)  # +/- 30% wind perturbation
        solar_factor = rng.uniform(0.80, 1.20)  # +/- 20% solar perturbation

        scenarios.append(
            {
                "load_factor": load_factor,
                "gen_factors": {
                    "baseload": baseload_factor,
                    "intermediate": intermediate_factor,
                    "peaker": peaker_factor,
                },
                "wind_profile": np.clip(wind_base * wind_factor, 0, 1),
                "solar_profile": np.clip(solar_base * solar_factor, 0, 1),
            }
        )
    return scenarios


def run(network_file: str = NETWORK_FILE) -> dict:
    """Execute stochastic wrapping test and return structured results."""
    results = {
        "status": "fail",
        "wall_clock_seconds": 0.0,
        "details": {},
        "errors": [],
        "workarounds": [],
    }

    start_total = time.perf_counter()
    try:
        # 1. Load base network
        n_base = _load_network_with_costs(network_file)

        # Set up 12 hourly snapshots
        snapshots = pd.date_range("2024-01-01", periods=N_HOURS, freq="h")
        n_base.set_snapshots(snapshots)

        # Set time-varying loads (constant for base, scaled per scenario)
        base_loads = n_base.loads["p_set"].copy()
        load_df = pd.DataFrame(
            {load_idx: base_loads[load_idx] for load_idx in n_base.loads.index},
            index=snapshots,
        )
        n_base.loads_t.p_set = load_df

        # Add renewables
        wind_base, solar_base = _add_renewables(n_base, snapshots)

        # Classify generators by resource type
        gen_types = _classify_generators(n_base)

        # 2. Generate 20 scenarios
        rng = np.random.default_rng(seed=RNG_SEED)
        scenarios = _generate_scenarios(rng, gen_types, wind_base, solar_base, N_SCENARIOS)

        # 3. Solve each scenario
        all_lmps = []
        all_dispatch = []
        all_objectives = []
        scenario_times = []

        results["workarounds"].append(
            "Manually set marginal_cost from gencost data (PPC importer does not import gencost)"
        )

        for s_idx, scenario in enumerate(scenarios):
            # Copy network for this scenario
            n = n_base.copy()

            # Apply load perturbation
            load_scenario = pd.DataFrame(
                {
                    load_idx: base_loads[load_idx] * scenario["load_factor"]
                    for load_idx in n.loads.index
                },
                index=snapshots,
            )
            n.loads_t.p_set = load_scenario

            # Apply generator p_nom perturbations by resource type
            for gen_idx in n.generators.index:
                if gen_idx in gen_types:
                    gtype = gen_types[gen_idx]
                    if gtype in scenario["gen_factors"]:
                        factor = scenario["gen_factors"][gtype]
                        p_nom = n_base.generators.loc[gen_idx, "p_nom"]
                        n.generators.loc[gen_idx, "p_nom"] = p_nom * factor

            # Apply renewable profiles for this scenario
            p_max_pu = pd.DataFrame(1.0, index=snapshots, columns=n.generators.index)
            p_max_pu["Wind_1"] = scenario["wind_profile"]
            p_max_pu["Solar_1"] = scenario["solar_profile"]
            n.generators_t.p_max_pu = p_max_pu

            # Solve DCOPF
            t0 = time.perf_counter()
            status = n.optimize(solver_name=SOLVER_NAME, solver_options=SOLVER_OPTIONS)
            t1 = time.perf_counter()
            scenario_times.append(t1 - t0)

            converged = "ok" in str(status).lower() or "optimal" in str(status).lower()
            if not converged:
                results["errors"].append(f"Scenario {s_idx} failed: {status}")
                continue

            # Collect results
            lmps = n.buses_t.marginal_price.copy()
            lmps["scenario"] = s_idx
            all_lmps.append(lmps)

            dispatch = n.generators_t.p.copy()
            dispatch["scenario"] = s_idx
            all_dispatch.append(dispatch)

            all_objectives.append(float(n.objective))

        elapsed_total = time.perf_counter() - start_total

        # 4. Aggregate results
        n_solved = len(all_objectives)
        if n_solved == 0:
            results["errors"].append("No scenarios converged")
            results["wall_clock_seconds"] = elapsed_total
            return results

        # Combine into structured DataFrames
        lmp_df = pd.concat(all_lmps, ignore_index=True)
        dispatch_df = pd.concat(all_dispatch, ignore_index=True)

        # Summary statistics
        obj_arr = np.array(all_objectives)
        time_arr = np.array(scenario_times)

        # LMP statistics across scenarios (excluding 'scenario' column)
        lmp_cols = [c for c in lmp_df.columns if c != "scenario"]
        lmp_values = lmp_df[lmp_cols].values

        # Dispatch statistics
        dispatch_cols = [c for c in dispatch_df.columns if c != "scenario"]
        dispatch_values = dispatch_df[dispatch_cols].values

        results["details"] = {
            "n_scenarios_requested": N_SCENARIOS,
            "n_scenarios_solved": n_solved,
            "n_hours": N_HOURS,
            "solver": SOLVER_NAME,
            "total_time_s": elapsed_total,
            "solve_time_total_s": float(time_arr.sum()),
            "solve_time_mean_s": float(time_arr.mean()),
            "solve_time_std_s": float(time_arr.std()),
            "solve_time_min_s": float(time_arr.min()),
            "solve_time_max_s": float(time_arr.max()),
            "objective_mean": float(obj_arr.mean()),
            "objective_std": float(obj_arr.std()),
            "objective_min": float(obj_arr.min()),
            "objective_max": float(obj_arr.max()),
            "lmp_mean": float(np.nanmean(lmp_values)),
            "lmp_std": float(np.nanstd(lmp_values)),
            "lmp_min": float(np.nanmin(lmp_values)),
            "lmp_max": float(np.nanmax(lmp_values)),
            "dispatch_total_mean_MW": float(dispatch_values.sum(axis=1).mean()),
            "result_format": "pandas DataFrame with scenario column",
            "api_mechanism": (
                "n.copy() for per-scenario network, "
                "direct DataFrame assignment for timeseries, "
                "n.optimize() per scenario"
            ),
            "generator_types": gen_types,
            "renewable_generators_added": ["Wind_1", "Solar_1"],
        }

        results["wall_clock_seconds"] = elapsed_total

        # Pass condition: all scenarios solved, results in structured format
        if n_solved == N_SCENARIOS:
            results["status"] = "pass"
        elif n_solved >= N_SCENARIOS * 0.9:
            results["status"] = "qualified_pass"

    except Exception as e:
        results["errors"].append(f"{type(e).__name__}: {e}")
        results["details"]["traceback"] = traceback.format_exc()
        results["wall_clock_seconds"] = time.perf_counter() - start_total

    return results


if __name__ == "__main__":
    import json

    result = run()
    print(json.dumps(result, indent=2, default=str))
